ft-sdk/src/auth/session.rs

```diff
@@ -4,56 +4,272 @@ pub struct SessionID(pub String);
+                .execute(conn)?;
+            Ok(create_with_user(
+                conn,
+                user_id,
+                session_expiration_duration,
+            )?)
+        }
+        Some((session_id, false)) => {
+            // Session is not expired, update the user ID
+            diesel::update(fastn_session::table.filter(fastn_session::id.eq(session_id.as_str())))
+                .set((
+                    fastn_session::uid.eq(Some(user_id.0)),
+                    fastn_session::updated_at.eq(now),
+                ))
+                .execute(conn)?;
+
+            Ok(SessionID(session_id))
+        }
+        None => {
+            // Session does not exist, create a new one
+            Ok(create_with_user(
+                conn,
+                user_id,
+                session_expiration_duration,
+            )?)
+        }
     }
 }
 
 #[cfg(feature = "auth-provider")]
 pub fn create_with_user(
     conn: &mut ft_sdk::Connection,
-    user_id: i64,
+    ft_sdk::UserId(user_id): &ft_sdk::UserId,
+    session_expiration_duration: Option<chrono::Duration>,
 ) -> Result<ft_sdk::auth::SessionID, diesel::result::Error> {
     use diesel::prelude::*;
     use ft_sdk::auth::fastn_session;
+    use std::ops::Add;
 
     let session_id = generate_new_session_id();
+    let session_expires_at =
+        session_expiration_duration.map(|duration| ft_sdk::env::now().add(duration));
 
     diesel::insert_into(fastn_session::table)
         .values((
             fastn_session::id.eq(&session_id),
-            fastn_session::uid.eq(Some(user_id)),
+            fastn_session::uid.eq(Some(*user_id)),
             fastn_session::created_at.eq(ft_sdk::env::now()),
             fastn_session::updated_at.eq(ft_sdk::env::now()),
+            fastn_session::expires_at.eq(session_expires_at),
             fastn_session::data.eq("{}"),
         ))
         .execute(conn)?;
 
     Ok(ft_sdk::auth::SessionID(session_id))
 }
 
+/// Sets a session cookie with an expiration time based on the session's expiration time
+/// in the database. If the session has an expiration time in the future, the cookie's
+/// max age is set to the remaining duration until that time. Otherwise, a default max age
+/// of 400 days is set.
+///
+/// # Arguments
+///
+/// * `conn` - A mutable reference to the database connection.
+/// * `session_id` - A string slice representing the session ID.
+/// * `host` - The host for which the cookie is valid.
+///
+/// # Errors
+///
+/// This function will return an error if:
+/// * The session ID is not found in the database.
+/// * The session ID is found but has expired.
+/// * There is an issue querying the database.
+/// * There is an error creating the `http::HeaderValue`.
+#[cfg(feature = "auth-provider")]
+pub fn set_session_cookie(
+    conn: &mut ft_sdk::Connection,
+    ft_sdk::auth::SessionID(session_id): ft_sdk::auth::SessionID,
+    host: ft_sdk::Host,
+) -> Result<http::HeaderValue, ft_sdk::Error> {
```

Comment:
Dont think this function should exist. Also if it existed this function must generate the session id, why would you call one function to get session id and then call another function to set session cookie, when one function will do.

Reply:
I need `session_id` separately to pass as access_token. Though, I see we can merge this.